Update the Asset Pipeline guide (#47012)
rails/rails: guides/source/asset_pipeline.md

The Asset Pipeline
==================

This guide covers the asset pipeline as it was implemented by Sprockets.

After reading this guide, you will know:

* What the asset pipeline is and what it does.
* How to properly organize your application assets.
* The benefits of the asset pipeline.
* How to package assets with a gem.

--------------------------------------------------------------------------------

What is the Asset Pipeline?
---------------------------

> The asset pipeline provides a framework to handle the delivery of Javascript and

Reviewer: ```suggestion
The asset pipeline provides a framework to handle the delivery of JavaScript and
```

What is the Asset Pipeline?
---------------------------

The asset pipeline provides a framework to handle the delivery of JavaScript and
CSS assets. This is done by leveraging technologies like HTTP/2 and techniques like
concatenation and minification. Finally, it allows your
application to be automatically combined with assets from other gems.

The asset pipeline is implemented by the [importmap-rails](https://github.com/rails/importmap-rails), [sprockets](https://github.com/rails/sprockets) and [sprockets-rails](https://github.com/rails/sprockets-rails) gems, and is enabled by default. You can disable it while creating a new application by
passing the `--skip-asset-pipeline` option.

```bash
$ rails new appname --skip-asset-pipeline
```

> NOTE: This guide is focusing on the default asset pipeline using only `Sprockets` for CSS and `Importmaps-rails` for JavaScript processing. The main limitation of those two is that there is no support for transpiling so you can't use things like `Babel`, `Typescript`, `Sass`, `React JSX format` or `TailwindCSS`. We encourage you to read the [Alternative Libraries section](#alternative-libraries) if you need transpiling for your JavaScript/CSS.

Reviewer 2: - `Importmaps-rails`-> `importmap-rails`

into each filename so that the file is cached by the web browser and CDN. This
fingerprint is automatically updated when you change the file contents, which
invalidates the cache.

The second feature of the asset pipeline is to use [import maps](https://github.com/WICG/import-maps)
when serving JavaScript files. This lets you build modern applications using
Javascript libraries made for ES modules (ESM) without the need for transpiling
and bundling. In turn, **this eliminates the need for Webpack, yarn, node or any
other part of the JavaScript toolchain**.

The third feature of the asset pipeline is to concatenate all CSS files into
one main `.css` file, which is then minified or compressed.
As you'll learn later in this guide, you can customize this strategy to group
files any way you like. In production, Rails inserts a SHA256 fingerprint into
each filename so that the file is cached by the web browser. You can invalidate
the cache by altering this fingerprint, which happens automatically whenever you
change the file contents.

The fourth feature of the asset pipeline is it allows coding assets via a
higher-level language for CSS.

> The first feature of the pipeline is to concatenate assets, which can reduce the

Reviewer 2: Looks like the four features are duplicate.

With this approach, you'll ship many small JavaScript files instead of one big JavaScript file. Thanks to HTTP/2 that no longer carries a material performance penalty during the initial transport, and in fact offers substantial benefits over the long run due to better caching dynamics.

How to use Import Maps as Javascript Asset Pipeline
-----------------------------

Import Maps are the default Javascript processor, the logic of generating import maps is handled by the [`importmap-rails`](https://github.com/rails/importmap-rails) gem.

WARNING: Importmaps are used only for Javascript files and can not be used for CSS delivery. Check the [Sprockets section](#how-to-use-sprockets) to learn about CSS.

You can find detailed usage instructions on the Gem homepage, but it's important to understand the basics of `importmap-rails`.

### How it works

Importmaps are essentially a string substitution for what is referred to as "bare module specifiers". A "bare module specifier" looks like this: 

```javascript
import React from "react"
```

> This is **not compatible** with the ES Module loader spec. Instead, **to be ESM compatible**, you must provide 1 of the 3 following types of specifiers:

Reviewer: I think for a first time user this is probably a bit technical, it's talking about concepts they may have never come across or don't care about.

For someone coming from a package.json world it's great, but I don't know if it needs to be the first thing you see here.

Author: I've updated this part to show import maps more like an "alias" for the library source address, as it's the simplest definition of an import map I could think of. What do you think about the newer version?

```

You can control the environment of the package for packages with separate "production" (the default) and "development" builds:

```sh
./bin/importmap pin react --env development
Pinning "react" to https://ga.jspm.io/npm:react@17.0.2/dev.index.js
Pinning "object-assign" to https://ga.jspm.io/npm:object-assign@4.1.1/index.js
```

You can also pick an alternative, supported CDN provider when pinning, like `unpkg` or `jsdelivr` (`jspm` is the default):

```sh
./bin/importmap pin react --from jsdelivr
Pinning "react" to https://cdn.jsdelivr.net/npm/react@17.0.2/index.js
```

Remember, though, that if you switch a pin from one provider to another, you may have to clean up dependencies added by the first provider that isn't used by the second provider.

Run `./bin/importmap` to see all options.

Note that this command is merely a convenience wrapper to resolving logical package names to CDN URLs. You can also just lookup the CDN URLs yourself, and then pin those. For example, if you wanted to use Skypack for React, you could just add the following to `config/importmap.rb`:

```ruby
pin "react", to: "https://cdn.skypack.dev/react"
```

### Preloading pinned modules

> To avoid the waterfall effect where the browser has to load one file after another before it can get to the deepest nested import, importmap-rails supports [modulepreload links](https://developers.google.com/web/updates/2017/12/modulepreload). Pinned modules can be preloaded by appending `preload: true` to the pin.

Reviewer: Is there some generic advice we can give on which modules you should/shouldn't preload?

Author: I did some Googling and I didn't find anything super useful, most clarifying for me was this answer:
https://stackoverflow.com/a/63924789/9984998

And I am thinking of adding such a sentence to the guide:
> Preloading is extremely useful in case you know that you'd need a specific part of the code, but you don't want to execute it immediately.

Getting to specific about the library might be difficult, what do you think?

Reviewer: How about this:

> It's a good idea to preload libraries or frameworks that are used throughout your app, as this will tell the browser to download them sooner.

Stimulus and Turbo get preloaded by default

https://github.com/hotwired/stimulus-rails/blob/498902a3687d32bd6589f7d3c7e7d72eabd822c0/lib/install/stimulus_with_importmap.rb#L15

And I think this explanation is okay for why.

Example:

```ruby
# config/importmap.rb
pin "@github/hotkey", to: "https://ga.jspm.io/npm:@github/hotkey@1.4.4/dist/index.js", preload: true
pin "md5", to: "https://cdn.jsdelivr.net/npm/md5@2.3.0/md5.js"

# app/views/layouts/application.html.erb
<%= javascript_importmap_tags %>

# will include the following link before the importmap is setup:
<link rel="modulepreload" href="https://ga.jspm.io/npm:@github/hotkey@1.4.4/dist/index.js">
...
```

NOTE: Refer to [`importmap-rails`](https://github.com/rails/importmap-rails) repository for the most up-to-date documentation.

How to Use Sprockets
-----------------------------

> In older versions of Rails, all assets were located in subdirectories of

Reviewer: I personally find it unhelpful when you read the current Rails guides and they say "older versions of Rails did X". Particularly when working on a new app, I don't care - I want to know the best way to do things now and into the future.

My preference is to have a note about upgrading/compatibility and have it marked as such for people who need to know.

Author: Rewritten trying to show why Sprockets is helpful and what kind of problems it solve:
> The naive approach to expose your application assets to the web would be to store them in
subdirectories of `public` folder such as `images` and `stylesheets`. Doing so manually would be difficult as most of the modern web applications require the assets to be processed in specific way for eg. compressing and adding fingerprints to the assets.

Sprockets is designed to automatically preprocess your assets stored in the configured directories and after processing expose them in the `public/assets` folder with fingerprinting, compression, source maps generation and other configurable features.

for files that must undergo some pre-processing before they are served.

In production, Rails precompiles these files to `public/assets` by default. The
precompiled copies are then served as static assets by the web server. The files
in the `app/assets` are never served directly in production.

[`config.public_file_server.enabled`]: configuring.html#config-public-file-server-enabled

### Manifest Files and Directives

When compiling assets with Sprockets, Sprockets needs to decide which top-level targets to compile, usually `application.css` and images. The top-level targets are defined in the Sprockets `manifest.js` file, by default it looks like this:

```js
//= link_tree ../images
//= link_directory ../stylesheets .css
//= link_tree ../../javascript .js
//= link_tree ../../../vendor/javascript .js
```

It contains _directives_ - instructions that tell Sprockets
which files to require in order to build a single CSS or JavaScript file.

This is meant to include the contents of all files found in the `./app/assets/images` directory or any subdirectories as well as any file recognized as JS directly at `./app/javascript` or `./vendor/javascript`.

> It will load any CSS from the `./app/assets/stylesheets` directory(not including subdirectories). Assuming that you have `application.css` and `marketing.css` files in the `./app/assets/stylesheets` folder it will allow you to load those stylesheets with `<%= stylesheet_link_tag "application" %>` or `<%= stylesheet_link_tag "marketing" %>` from your views.

Reviewer: ```suggestion
It will load any CSS from the `./app/assets/stylesheets` directory (not including subdirectories). Assuming that you have `application.css` and `marketing.css` files in the `./app/assets/stylesheets` folder it will allow you to load those stylesheets with `<%= stylesheet_link_tag "application" %>` or `<%= stylesheet_link_tag "marketing" %>` from your views.
```

global-908e25f4bf641868d8683022a5b62f54.css
```

Fingerprinting is enabled by default for both the development and production
environments. You can enable or disable it in your configuration through the
[`config.assets.digest`][] option.

### What are Import Maps and Why Should I Care?

Import maps let you import JavaScript modules using logical names that map to versioned/digested files – directly from the browser. So you can build modern JavaScript applications using JavaScript libraries made for ES modules (ESM) without the need for transpiling or bundling.

With this approach, you'll ship many small JavaScript files instead of one big JavaScript file. Thanks to HTTP/2 that no longer carries a material performance penalty during the initial transport, and in fact offers substantial benefits over the long run due to better caching dynamics.

How to use Import Maps as Javascript Asset Pipeline
-----------------------------

Import Maps are the default Javascript processor, the logic of generating import maps is handled by the [`importmap-rails`](https://github.com/rails/importmap-rails) gem.

> WARNING: Importmaps are used only for Javascript files and can not be used for CSS delivery. Check the [Sprockets section](#how-to-use-sprockets) to learn about CSS.

Reviewer 2: Importmaps -> Import maps

You can find detailed usage instructions on the Gem homepage, but it's important to understand the basics of `importmap-rails`.

### How it works

> Importmaps are essentially a string substitution for what is referred to as "bare module specifiers". They allow you to standardize the names of JavaScript module imports.

Reviewer 2: Importmaps -> import maps

Take for example such an import definition, it won't work without an import map:

```javascript
import React from "react"
```

You would have to define it like this to make it work:

```javascript
import React from "https://ga.jspm.io/npm:react@17.0.2/index.js"
```

Here comes the import map, we define the `react` name to be pinned to the `https://ga.jspm.io/npm:react@17.0.2/index.js` address. With such information our browser accepts the simplified `import React from "react"` definition. Think of import map as about an alias for the library source address.

### Usage

With `importmap-rails` you create the importmap configuration file pinning the library path to a name:

```ruby
# config/importmap.rb
pin "application"
pin "react", to: "https://ga.jspm.io/npm:react@17.0.2/index.js"
```

> All of the configured import maps should be attached in `<head>` tag of your application by adding  `<%= javascript_importmap_tags %>` . The `javascript_importmap_tags` renders a bunch of scripts in the `head` tag:

Reviewer 2: - `<head>` tag -> `<head>` element
- `head` tag -> `<head>` element

- JSON with all configured import maps:

```html
<script type="importmap">
{
  "imports": {
    "application": "/assets/application-39f16dc3f3....js"
    "react": "https://ga.jspm.io/npm:react@17.0.2/index.js"
  }
}
</script>
```

- [`Es-module-shims`](https://github.com/guybedford/es-module-shims) acting as polyfill ensuring support for `import maps` on older browsers:

```html
<script src="/assets/es-module-shims.min" async="async" data-turbo-track="reload"></script>
```

> - Entrypoint for loading Javascript from `app/javascript/application.js`:

Reviewer 2: Javascript -> JavaScript

```html
<script type="module">import "application"</script>
```

### Using npm packages via JavaScript CDNs

> You can use the `./bin/importmap` command that's added as part of the `importmap-rails` install to pin, unpin, or update npm packages in your import map. The binstub uses `JSPM.org`

Reviewer 2: Just for convenience: `JSPM.org` -> [`JSPM.org`](https://jspm.org/)

    "react-dom": "https://ga.jspm.io/npm:react-dom@17.0.2/index.js",
    "object-assign": "https://ga.jspm.io/npm:object-assign@4.1.1/index.js",
    "scheduler": "https://ga.jspm.io/npm:scheduler@0.20.2/index.js"
  }
}
```

As you can see, the two packages react and react-dom resolve to a total of four dependencies, when resolved via the jspm default.

Now you can use these in your `application.js` entrypoint like you would any other module:

```javascript
import React from "react"
import ReactDOM from "react-dom"
```

You can also designate a specific version to pin:

```sh
./bin/importmap pin react@17.0.1
Pinning "react" to https://ga.jspm.io/npm:react@17.0.1/index.js
Pinning "object-assign" to https://ga.jspm.io/npm:object-assign@4.1.1/index.js
```

Or even remove pins:

```sh
./bin/importmap unpin react
Unpinning "react"
Unpinning "object-assign"
```

You can control the environment of the package for packages with separate "production" (the default) and "development" builds:

```sh
./bin/importmap pin react --env development
Pinning "react" to https://ga.jspm.io/npm:react@17.0.2/dev.index.js
Pinning "object-assign" to https://ga.jspm.io/npm:object-assign@4.1.1/index.js
```

> You can also pick an alternative, supported CDN provider when pinning, like `unpkg` or `jsdelivr` (`jspm` is the default):

Reviewer 2: Just for convenience: 
`unpkg` -> [`unpkg`](https://unpkg.com/)
`jsdelivr` -> [`jsdelivr`](https://www.jsdelivr.com/)

In the `manifest.js` you could also specify the `link` directive to load a specific file instead of the whole directory. `link` directive requires providing explicit file extension.

Sprockets load the files specified, processes them if
necessary, concatenates them into one single file, and then compresses them
(based on the value of `config.assets.css_compressor` or `config.assets.js_compressor`).
Compression reduces file size, enabling the browser to download the files faster.

### Controller Specific Assets

When you generate a scaffold or a controller, Rails also generates a
Cascading Style Sheet file for that controller. Additionally, when generating a scaffold, Rails generates the file `scaffolds.css`.

For example, if you generate a `ProjectsController`, Rails will also add a new
file at `app/assets/stylesheets/projects.css`. By default, these files will be
ready to use by your application immediately using the `link_directory` directive in the `manifest.js` file.

You can also opt to include controller-specific stylesheets files
only in their respective controllers using the following:

> `<%= stylesheet_link_tag params[:controller] %>`

Reviewer 2: Wrapping with three backquotes with `html+erb` is preferable:

```html+erb
<%= stylesheet_link_tag params[:controller] %>
```

### [jsbundling-rails](https://github.com/rails/jsbundling-rails)

A Node.js dependent alternative to the `importmap-rails` way of bundling JavaScript with [esbuild](https://esbuild.github.io/), [rollup.js](https://rollupjs.org/), or [Webpack](https://webpack.js.org/).

The gem provides `yarn build --watch` process to automatically generate output in development. For production, it automatically hooks `javascript:build` task into `assets:precompile` task to ensure that all your package dependencies have been installed and JavaScript has been built for all entry points.

**When to use instead of `importmap-rails`?** If your JavaScript code depends on transpiling so if you are using [Babel](https://babeljs.io/), [TypeScript](https://www.typescriptlang.org/) or React `JSX` format then `jsbundling-rails` is the correct way to go.

### [Webpacker/Shakapacker](webpacker.html)

Webpacker was the default JavaScript pre-processor and bundler for Rails 5 and 6. It has now been retired. A successor called [`shakapacker`](https://github.com/shakacode/shakapacker) exists, but is not maintained by the Rails team or project.

> Unlike other libraries in this list `webpacker`/`shakapacker` is completely independent of Sprockets and could process both JavaScript and CSS files. Read the [Webpacker guide](https://guides.rubyonrails.org/webpacker.html) to learn more.

Reviewer 2: https://guides.rubyonrails.org/webpacker.html -> webpacker.html

We are aware that there are no one-size-fits-it-all solutions for the various JavaScript and CSS frameworks/extensions available. There are other bundling libraries in the Rails ecosystem that should empower you in the cases where the default setup isn't enough.

### [jsbundling-rails](https://github.com/rails/jsbundling-rails)

A Node.js dependent alternative to the `importmap-rails` way of bundling JavaScript with [esbuild](https://esbuild.github.io/), [rollup.js](https://rollupjs.org/), or [Webpack](https://webpack.js.org/).

The gem provides `yarn build --watch` process to automatically generate output in development. For production, it automatically hooks `javascript:build` task into `assets:precompile` task to ensure that all your package dependencies have been installed and JavaScript has been built for all entry points.

**When to use instead of `importmap-rails`?** If your JavaScript code depends on transpiling so if you are using [Babel](https://babeljs.io/), [TypeScript](https://www.typescriptlang.org/) or React `JSX` format then `jsbundling-rails` is the correct way to go.

### [Webpacker/Shakapacker](https://guides.rubyonrails.org/webpacker.html)

> Webpacker has been the default JavaScript pre-processor and bundler for Rails 5 and 6. `Webpacker` successor is now [`shakapacker`](https://github.com/shakacode/shakapacker) and it's advised to use it.

Reviewer: ```suggestion
Webpacker was the default JavaScript pre-processor and bundler for Rails 5 and 6. It has now been retired. A successor called [`shakapacker`](https://github.com/shakacode/shakapacker) exists, but is not maintained by the Rails team or project.
```

Author: The first line of README in Shakapacker repo goes like this:
> Official, actively maintained successor to [rails/webpacker](https://github.com/rails/webpacker).

If it's not maintained by Rails team then maybe they shouldn't use the `official` word, but anyway I like the rewording there and going to commit it.

Reviewer: I'm not sure what official means in that context. It was added in https://github.com/shakacode/shakapacker/pull/14

I suggested that wording because it's not part of the `rails/` github org so isn't owned by the same core team.

cc @justin808 you may have some thoughts on this (or on this PR in general).

Reviewer 3: It's official in the sense that @dhh recommends it as the successor, and he came up with the name "shakapacker."

See https://github.com/rails/webpacker.

> The `webpacker`/`shakapacker` unlike other libraries in this list is completely independent of Sprockets and could process both JavaScript and CSS files. We encourage you to read the [Webpacker guide](https://guides.rubyonrails.org/webpacker.html) to learn more.

Reviewer: ```suggestion
Unlike other libraries in this list, `webpacker`/`shakapacker` is completely independent of Sprockets and could process both JavaScript and CSS files. Read the [Webpacker guide](https://guides.rubyonrails.org/webpacker.html) to learn more.
```

NOTE: Read the [Comparison with Webpacker](https://github.com/rails/jsbundling-rails/blob/main/docs/comparison_with_webpacker.md) document to understand the differences between `jsbundling-rails` and `webpacker`/`shakapacker`.

***

### [cssbundling-rails](https://github.com/rails/cssbundling-rails)

Allows bundling and processing of your CSS using [Tailwind CSS](https://tailwindcss.com/), [Bootstrap](https://getbootstrap.com/), [Bulma](https://bulma.io/), [PostCSS](https://postcss.org/), or [Dart Sass](https://sass-lang.com/), then delivers the CSS via the asset pipeline.

It works in a similar way to `jsbundling-rails` so adds the Node.js dependency to your application with `yarn build:css --watch` process to regenerate your stylesheets in development and hooks into `assets:precompile` task in production.

**What's the difference between Sprockets?** Sprockets on their own is not able to transpile the Sass into CSS, Node.js is required to generate the `.css` files from your `.sass`  files. Once the `.css` files are generated then `Sprockets` is able to deliver them to your clients.

> NOTE: `cssbundling-rails` is relying on Node to process the CSS. The `dartsass-rails` and `tailwindcss-rails` gems are using the standalone versions of Tailwind CSS and Dart Sass meaning no Node dependency. If you are using `importmap-rails` to handle your Javascripts and `dartsass-rails` or `tailwindcss-rails` for CSS you could completely avoid the Node dependency resulting in a less complex solution.

Reviewer: ```suggestion
NOTE: `cssbundling-rails` relies on Node to process CSS. The `dartsass-rails` and `tailwindcss-rails` gems use standalone versions of Tailwind CSS and Dart Sass, meaning no Node dependency. If you are using `importmap-rails` to handle your Javascripts and `dartsass-rails` or `tailwindcss-rails` for CSS you could completely avoid the Node dependency resulting in a less complex solution.
```

### [dartsass-rails](https://github.com/rails/dartsass-rails)

> If you are willing to use [`Sass`](https://sass-lang.com/) in your application `dartsass-rails` comes as a replacement for the legacy `sassc-rails` gem. `dartsass-rails` is using the `Dart Sass` implementation in favour of deprecated in 2020 [`LibSass`](https://sass-lang.com/blog/libsass-is-deprecated) used by `sassc-rails`.

Reviewer: ```suggestion
If you want to use [`Sass`](https://sass-lang.com/) in your application, `dartsass-rails` comes as a replacement for the legacy `sassc-rails` gem. `dartsass-rails` uses the `Dart Sass` implementation in favour of deprecated in 2020 [`LibSass`](https://sass-lang.com/blog/libsass-is-deprecated) used by `sassc-rails`.
```

We are aware that there are no one-size-fits-it-all solutions for the various JavaScript and CSS frameworks/extensions available. There are other bundling libraries in the Rails ecosystem that should empower you in the cases where the default setup isn't enough.

### [jsbundling-rails](https://github.com/rails/jsbundling-rails)

A Node.js dependent alternative to the `importmap-rails` way of bundling JavaScript with [esbuild](https://esbuild.github.io/), [rollup.js](https://rollupjs.org/), or [Webpack](https://webpack.js.org/).

The gem provides `yarn build --watch` process to automatically generate output in development. For production, it automatically hooks `javascript:build` task into `assets:precompile` task to ensure that all your package dependencies have been installed and JavaScript has been built for all entry points.

**When to use instead of `importmap-rails`?** If your JavaScript code depends on transpiling so if you are using [Babel](https://babeljs.io/), [TypeScript](https://www.typescriptlang.org/) or React `JSX` format then `jsbundling-rails` is the correct way to go.

> ### [Webpacker/Shakapacker](https://guides.rubyonrails.org/webpacker.html)

Reviewer 2: FYI: while the situations and prospects are still very very unclear, it seems that webpack is being moving to the new Turbopack: 
https://vercel.com/blog/the-turbopack-vision
